idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/AddWhenRemainingBranchFixFactories.kt

```diff
@@ -0,0 +1,76 @@
+import org.jetbrains.kotlin.idea.KotlinBundle
+import org.jetbrains.kotlin.idea.api.applicator.HLApplicator
+import org.jetbrains.kotlin.idea.api.applicator.HLApplicatorInput
+import org.jetbrains.kotlin.idea.api.applicator.applicator
+import org.jetbrains.kotlin.idea.fir.api.fixes.HLQuickFix
+import org.jetbrains.kotlin.idea.fir.api.fixes.diagnosticFixFactory
+import org.jetbrains.kotlin.idea.frontend.api.analyse
+import org.jetbrains.kotlin.idea.frontend.api.components.ShortenOption
+import org.jetbrains.kotlin.idea.frontend.api.fir.diagnostics.KtFirDiagnostic
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtClassKind
+import org.jetbrains.kotlin.idea.frontend.api.tokens.HackToForceAllowRunningAnalyzeOnEDT
+import org.jetbrains.kotlin.idea.frontend.api.tokens.hackyAllowRunningOnEdt
+import org.jetbrains.kotlin.idea.util.generateWhenBranches
+import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.psi.KtWhenExpression
+
+object AddWhenRemainingBranchFixFactories {
+    class Input(val whenMissingCases: List<WhenMissingCase>, val enumToStarImport: ClassId?) : HLApplicatorInput
+
+    val applicator: HLApplicator<KtWhenExpression, Input> = getApplicator(false)
+    val applicatorUsingStarImport: HLApplicator<KtWhenExpression, Input> = getApplicator(true)
+
+    @OptIn(HackToForceAllowRunningAnalyzeOnEDT::class)
+    private fun getApplicator(useStarImport: Boolean = false) = applicator<KtWhenExpression, Input> {
+        familyAndActionName(
+            if (useStarImport) KotlinBundle.lazyMessage("fix.add.remaining.branches.with.star.import")
+            else KotlinBundle.lazyMessage("fix.add.remaining.branches")
+        )
+        applyTo { whenExpression, input ->
+            if (useStarImport) assert(input.enumToStarImport != null)
+            generateWhenBranches(whenExpression, input.whenMissingCases)
+            val shortenCommand = hackyAllowRunningOnEdt {
+                analyse(whenExpression) {
+                    collectPossibleReferenceShorteningsInElement(
+                        whenExpression,
+                        callableShortenOption = {
+                            if (useStarImport && it.callableIdIfNonLocal?.classId == input.enumToStarImport) {
+                                ShortenOption.SHORTEN_AND_IMPORT_ALL_IN_PARENT
+                            } else {
+                                ShortenOption.DO_NOT_SHORTEN
+                            }
+                        })
+
+                }
+            }
+            shortenCommand.invokeShortening()
+        }
+    }
+
+    @OptIn(ExperimentalStdlibApi::class)
+    val noElseInWhen = diagnosticFixFactory(KtFirDiagnostic.NoElseInWhen::class) { diagnostic ->
+        val whenExpression = diagnostic.psi
+        val subjectExpression = whenExpression.subjectExpression ?: return@diagnosticFixFactory emptyList()
+
+        buildList {
+            val missingCases = whenExpression.getMissingCases().takeIf {
```

Comment:
Why don't we get missing cases from `NoElseInWhen` diagnosic? it contains `NoElseInWhen.missingWhenCases` property

Reply:
Oh... I missed that...
So do you think if it's worth to keeping the high level API to create missing cases from a when expression? Maybe it can be useful for a future intention to generate them for non-sealed when expressions (since there probably won't be any diagnostics for it)?

Reply:
Sure, let's keep it. But we should check that it will work for when used as non-expression. Mind adding tests that check that?

Reply:
OK, added HLAddWhenRemainingBranchesIntention in the last commit and enabled the corresponding intention test.